src/backend/distributed/commands/multi_copy.c

```diff
@@ -2923,3 +2984,216 @@ CopyGetAttnums(TupleDesc tupDesc, Relation rel, List *attnamelist)
+		placementState->shardState = shardState;
+		placementState->data = makeStringInfo();
+		placementState->connectionState = connectionState;
+		dlist_push_head(&connectionState->bufferedPlacementList,
+						&placementState->bufferedPlacementNode);
+		shardState->placementStateList = lappend(shardState->placementStateList,
+												 placementState);
+	}
+
+	shardState->shardId = shardId;
+}
+
+
+static MultiConnection *
+CopyGetPlacementConnection(ShardPlacement *placement)
+{
+	MultiConnection *connection = NULL;
+	uint32 connectionFlags = FOR_DML;
+	char *nodeUser = CurrentUserName();
+
+	/*
+	 ********************************************************
+	 * TODO: This is not a correct implementation, this should be implemented in a
+	 * compatible way with executor.c.
+	 * ******************************************************
+	 */
+
+	/*
+	 * Make sure we use a separate connection per placement for hash-distributed
+	 * tables in order to allow multi-shard modifications in the same transaction.
+	 */
+	if (placement->partitionMethod == DISTRIBUTE_BY_HASH)
+	{
+		connectionFlags |= CONNECTION_PER_PLACEMENT;
+	}
+
+	connection = GetPlacementConnection(connectionFlags, placement, nodeUser);
+
+	if (PQstatus(connection->pgConn) != CONNECTION_OK)
+	{
+		/* TODO: make this like OpenCopyConnections */
+		ReportConnectionError(connection, ERROR);
+	}
+
+	/*
+	 * Errors are supposed to cause immediate aborts (i.e. we don't
+	 * want to/can't invalidate placements), mark the connection as
+	 * critical so later errors cause failures.
+	 */
+	MarkRemoteTransactionCritical(connection);
+	ClaimConnectionExclusively(connection);
+	RemoteTransactionBeginIfNecessary(connection);
+
+	return connection;
+}
+
+
+static void
+StartPlacementStateCopyCommand(CopyPlacementState *placementState,
+							   CopyStmt *copyStatement, CopyOutState copyOutState,
+							   bool binaryCopy)
+{
+	StringInfo copyCommand = NULL;
+	PGresult *result = NULL;
+	MultiConnection *connection = placementState->connectionState->connection;
+	uint64 shardId = placementState->shardState->shardId;
+	bool raiseInterrupts = true;
+
+	copyCommand = ConstructCopyStatement(copyStatement, shardId, binaryCopy);
+
+	if (!SendRemoteCommand(connection, copyCommand->data))
+	{
+		ReportConnectionError(connection, ERROR);
+	}
+	result = GetRemoteCommandResult(connection, raiseInterrupts);
+	if (PQresultStatus(result) != PGRES_COPY_IN)
+	{
+		ReportResultError(connection, result, ERROR);
+	}
+	PQclear(result);
+
+	if (binaryCopy)
+	{
+		SendCopyBinaryHeaders(copyOutState, shardId, list_make1(connection));
+	}
+}
+
+
+static void
+EndPlacementStateCopyCommand(CopyPlacementState *placementState,
+							 CopyOutState copyOutState, bool binaryCopy)
+{
+	MultiConnection *connection = placementState->connectionState->connection;
+	uint64 shardId = placementState->shardState->shardId;
+
+	/* flush buffered data */
+	SendCopyDataToPlacement(placementState->data, shardId, connection);
```

Comment:
We're not guaranteed to flush all the data that is buffered in the PGconn 's`outBuffer` here because we're using non-blocking sockets. We should be calling FinishConnectionIO to flush all the data, otherwise `pqEndcopy3` will fail.

Reply:
The following `EndRemoteCopy()` calls `PutRemoteCopyEnd()` which calls `FinishConnectionIO()`.